### PR TITLE
uefi-dbx: Check only the efi executables from boot menu

### PR DESCRIPTION
**Pull request by hughsie**
Inspired from a patch by Crag Wang <crag.wang@dell.com>, many thanks.

Type of pull request:

- [ ] New plugin (Please include [new plugin checklist](https://github.com/fwupd/fwupd/wiki/New-plugin-checklist))
- [ ] Code fix
- [ ] Feature
- [ ] Documentation

### CODE REVIEW

File: plugins/uefi-dbx/fu-uefi-dbx-common.c [hunk 1]
```diff
@@ -29,13 +29,86 @@ fu_uefi_dbx_get_authenticode_hash(const gchar *fn, GError **error)
...
 
+static GPtrArray *
+fu_uefi_dbx_get_basenames_bootxxxx(void)
+{
+	g_autoptr(GPtrArray) files = g_ptr_array_new_with_free_func(g_free);
+
+	/* hardcoded, as we chainload from shim/grub to these */
+	g_ptr_array_add(files, g_strdup("fwupdx64.efi"));
+	g_ptr_array_add(files, g_strdup("fwupdaa64.efi"));
+	g_ptr_array_add(files, g_strdup("fwupdia32.efi"));
```
Comment (superm1) [inline plugins/uefi-dbx/fu-uefi-dbx-common.c:40]: we know the arch we're running on via `EFI_MACHINE_TYPE_NAME`, use that to build the string.

**Comment by hughsie** [comment]
Right, also my worry. The current code in main just looks at *everything* and that's never going to miss anything. This code would miss anything being chainloaded-to. I assume shim is also respecting the dbx too.

**Comment by superm1** [comment]
> Right, also my worry. The current code in main just looks at _everything_ and that's never going to miss anything. This code would miss anything being chainloaded-to. I assume shim is also respecting the dbx too.

I say build a heuristic of the "common" stuff that is chainloaded.  By looking at that heuristic and all the `BootXXXX` entries you'll get enough stuff I think to prevent common problems.

**Comment by hughsie** [comment]
> I say build a heuristic of the "common" stuff that is chainloaded

My worry is that it makes something that was very definite now very fuzzy.

**Comment by hughsie** [comment]
> parse grub.cfg

That just gave me the shivers. :)

**Comment by superm1** [comment]
> > parse grub.cfg
> 
> That just gave me the shivers. :)

Of course it could be crazy complex but at the end of the day it's really just look for a line with `chainload` in it and then any strings on that line match against files on the partition.

**Comment by hughsie** [comment]
I can see downsides with this patch -- and not a lot of upsides. Is the bug we're really trying to fix basically "ignore stuff left in the ESP from previous Linux installs"

**Comment by superm1** [comment]
> I can see downsides with this patch -- and not a lot of upsides. Is the bug we're really trying to fix basically "ignore stuff left in the ESP from previous Linux installs"

How about we add a config option to ignore selected files in `/etc/fwupd/daemon.conf`

**Comment by hughsie** [comment]
> How about we add a config option to ignore selected files in /etc/fwupd/daemon.conf

I don't think that helps much; if it's an old no-longer used EFI binary the user is better off just deleting it.

**Comment by superm1** [comment]
OK - then I think the remaining options are either:
1. Offer a config option to enable this PR behavior (only match executables in boot entries and extra special case ones - ignore chainloading)
2. Cancel this PR.

I personally lean on the second.  I think it's safer for anything we can't envision about how executables are used.

**Comment by hughsie** [comment]
> Offer a config option to enable this PR behavior

My trouble with this is -- why would the user **need** to ignore something? I guess it's less scary to set something in a config file than to delete something you don't know.

File: plugins/uefi-dbx/fu-uefi-dbx-common.c (hunk 1 above)
Comment (hughsie) [reply plugins/uefi-dbx/fu-uefi-dbx-common.c:40]: Fixed.

**Comment by hughsie** [comment]
Okay, I think a great solution here is to use --force to "downgrade" the checking. How about this?

File: plugins/uefi-dbx/fu-uefi-dbx-common.c [hunk 2]
```diff
@@ -29,13 +29,85 @@ fu_uefi_dbx_get_authenticode_hash(const gchar *fn, GError **error)
 	return g_strdup(fu_efi_image_get_checksum(img));
 }
 
+static GPtrArray *
+fu_uefi_dbx_get_basenames_bootxxxx(void)
+{
+	g_autoptr(GPtrArray) files = g_ptr_array_new_with_free_func(g_free);
+
+	/* hardcoded, as we chainload from shim/grub to these */
+	g_ptr_array_add(files, g_strdup_printf("fwupd%s.efi", EFI_MACHINE_TYPE_NAME));
```
Comment (superm1) [inline plugins/uefi-dbx/fu-uefi-dbx-common.c:38]: maybe we should also hardcode the grub prefixed ones because shim chainloads too.

**Review by superm1 — APPROVED** [review]
I think the new approach makes a lot of sense.

File: plugins/uefi-dbx/fu-uefi-dbx-common.c (hunk 2 above)
Comment (hughsie) [reply plugins/uefi-dbx/fu-uefi-dbx-common.c:38]: good idea, added.

**Review by superm1 — APPROVED** [review]

**Pull request merged**